File: ib_flex2ledger.rb
#!/usr/bin/env ruby
# typed: true

require 'rubygems'
require 'nokogiri'
require 'commander/import'
require 'pp'
require 'net/http'
require 'uri'
require 'csv'

program :name, 'ib_flex2ledger'
program :version, '0.1'
program :description, 'Parse Interactive Brokers Flex reports into ledger transactions'

def load(file)
  File.open(file) do |f|
    Nokogiri::XML(f)
  end
end

def transaction_header(trade)
  puts "#{trade["settleDateTarget"]}=#{trade["tradeDate"]} * #{trade["exchange"]}"
  puts "  ; trade_id: #{trade["tradeID"]}"
  puts "  ; order_id: #{trade["ibOrderID"]}"
end

def fees_to_ledger(trade, options)
  puts "  #{options.cash_account}  #{trade["ibCommissionCurrency"]} #{trade["ibCommission"].to_f}"
  puts "  #{options.fees_account}  #{trade["ibCommissionCurrency"]} #{-trade["ibCommission"].to_f}"
end

def stock_trade_to_ledger(trade, options)
  transaction_header(trade)
  puts "  #{options.stock_account}  \"#{trade["symbol"]}\" #{trade["quantity"].to_f}"
  puts "  #{options.cash_account}  #{trade["currency"]} #{trade["proceeds"].to_f}"
  fees_to_ledger(trade, options)
  puts ""
end

def fx_trade_to_ledger(trade, options)
  base_currency, quote_currency = trade["symbol"].split('.')

  transaction_header(trade)
  puts "  #{options.cash_account}  #{base_currency} #{trade["quantity"].to_f}"
  puts "  #{options.cash_account}  #{quote_currency} #{trade["proceeds"].to_f}"
  fees_to_ledger(trade, options)
  puts ""
end

def group_cash_transactions(cash_transactions)
  cash_transactions.reduce({}) do |hash, transaction|
    key = {conid: transaction["conid"], date: transaction["dateTime"]}

    if hash[key].nil? then
      hash[key] = [transaction]
    else
      hash[key] << transaction
    end

    hash
  end
end

def classify_cash_transactions_group(transaction_group)
  grouped = transaction_group.group_by {|transaction| transaction["type"]}

  results = []

  if grouped.fetch("Dividends",[]).length == 1 and grouped.fetch("Withholding Tax",[]).length == 1 then
    results << {type: :dividend_with_withholding, dividend: grouped["Dividends"].first, tax: grouped["Withholding Tax"].first}
    grouped.delete("Dividends")
    grouped.delete("Withholding Tax")
  end

  grouped.each do |key, transactions|
    transactions.each do |transaction|
      results << {type: key.to_sym, transaction: transaction}
    end
  end

  results
end

def retrieve_flex(api_token, query_id, wait_seconds)
  STDERR.print "Executing statement generation..."
  send_request_parameters = {
    'v': 3,
    't': api_token,
    'q': query_id,
  }

  send_request_uri = URI.parse('https://ndcdyn.interactivebrokers.com/AccountManagement/FlexWebService/SendRequest')
  send_request_uri.query = URI.encode_www_form(send_request_parameters)
  send_request_response = Net::HTTP.get_response(send_request_uri)
  send_request_result = Nokogiri::XML(send_request_response.body)

  unless send_request_result.xpath('//FlexStatementResponse/Status').text == 'Success' then
    raise RuntimeError.new("Response to /SendRequest was not Success:\n#{send_request_response.body}")
  end

  reference_code = send_request_result.xpath('//FlexStatementResponse/ReferenceCode').text
  STDERR.puts "done: #{reference_code}"

  STDERR.print "Waiting #{wait_seconds} seconds for statement to finish generating..."
  sleep(20)
  STDERR.puts "done"

  STDERR.print "Retrieving generated statement..."
  get_statement_parameters = {
    'v': 3,
    't': api_token,
    'q': reference_code,
  }

  get_statement_uri = URI.parse('https://ndcdyn.interactivebrokers.com/AccountManagement/FlexWebService/GetStatement')
  get_statement_uri.query = URI.encode_www_form(get_statement_parameters)
  get_statement_response = Net::HTTP.get_response(get_statement_uri)

  STDERR.puts("done")

  return get_statement_response.body
end

def parse_trades_from_flex(flex_report, options)
  statement = flex_report.xpath("//FlexStatement").first
  account_info = statement.xpath("AccountInformation").first
  $stderr.puts "Trades for #{account_info["name"]}, account #{statement["accountId"]}"
  $stderr.puts "Period #{statement["fromDate"]} to #{statement["toDate"]}"

  if options.new_only
    latest_transaction_date = get_latest_transaction_date_from_hledger(options.stock_account)
    $stderr.puts "Dropping transactions older than #{latest_transaction_date}"
  else
    latest_transaction_date = DateTime.new()
  end

  trades = statement.xpath("Trades/Trade")
  trades.sort_by {|trade| DateTime.parse("#{trade["tradeDate"]} #{trade["tradeTime"]}")}.each do |trade|
    next if DateTime.parse(trade["tradeDate"]) <= latest_transaction_date

    case trade["assetCategory"]
    when "CASH" then fx_trade_to_ledger(trade, options)
    when "STK"  then stock_trade_to_ledger(trade, options)
    else
      say_error "Dropping trade with unknown assetCategory=#{trade["assetCategory"]}"
      say_error trade.inspect
    end
  end

  cash_transactions = group_cash_transactions(statement.xpath("CashTransactions/CashTransaction[@levelOfDetail='DETAIL']"))
  cash_transactions.sort_by {|key, _| key[:date]}.each do |key, transactions|
    next if DateTime.parse(key[:date]) <= latest_transaction_date

    grouped_transactions = classify_cash_transactions_group(transactions)
    grouped_transactions.each do |transaction|
      case transaction[:type]
      when :dividend_with_withholding
        dividend = transaction[:dividend]
        tax = transaction[:tax]

        puts "#{dividend["reportDate"]} * #{dividend["symbol"]}"
        puts "  ; #{dividend["description"]}"
        puts "  #{options.dividends_account}  #{dividend["currency"]} #{-dividend["amount"].to_f}"
        puts "  #{options.withholdings_account}  #{tax["currency"]} #{-tax["amount"].to_f}"
        puts "  #{options.cash_account}"
        puts ""
      when :Dividends
        dividend = transaction[:transaction]

        puts "#{dividend["reportDate"]} * #{dividend["symbol"]}"
        puts "  ; #{dividend["description"]}"
        puts "  #{options.dividends_account}  #{dividend["currency"]} #{-dividend["amount"].to_f}"
        puts "  #{options.cash_account}"
        puts ""
      when :"Broker Interest Received"
        transaction = transaction[:transaction]

        puts "#{transaction["reportDate"]} * Interactive Brokers"
        puts "  ; #{transaction["description"]}"
        puts "  #{options.interest_income_account}  #{transaction["currency"]} #{-transaction["amount"].to_f}"
        puts "  #{options.cash_account}"
        puts ""
      when :"Broker Interest Paid"
        transaction = transaction[:transaction]

        puts "#{transaction["reportDate"]} * Interactive Brokers"
        puts "  ; #{transaction["description"]}"
        puts "  #{options.interest_expense_account}  #{transaction["currency"]} #{-transaction["amount"].to_f}"
        puts "  #{options.cash_account}"
        puts ""
      when :"Other Fees"
        transaction = transaction[:transaction]

        puts "#{transaction["reportDate"]} * Interactive Brokers"
        puts "  ; #{transaction["description"]}"
        puts "  #{options.fees_account}  #{transaction["currency"]} #{-transaction["amount"].to_f}"
        puts "  #{options.cash_account}"
        puts ""
      when :"Deposits/Withdrawals"
        next if options.ignore_deposits_withdrawals
        transaction = transaction[:transaction]

        puts "#{transaction["reportDate"]} * UNKNOWN"
        puts "  ; #{transaction["description"]}"
        puts "  #{options.cash_account}  #{transaction["currency"]} #{transaction["amount"].to_f}"
        puts "  UNKNOWN_ACCOUNT"
        puts ""
      else
        transaction = transaction[:transaction]
        amount = transaction["amount"].to_f

        puts "#{transaction["reportDate"]} * Interactive Brokers"
        puts "  ; #{transaction["description"]}"
        puts "  ; cash_transaction_type: #{transaction["type"]}"
        puts "  #{options.cash_account}  #{transaction["currency"]} #{transaction["amount"].to_f}"
        if amount < 0 then
          puts "  #{options.fees_account}"
        else
          puts "  UNKNOWN_ACCOUNT"
        end
        puts ""
      end
    end
  end
end

def get_latest_transaction_date_from_hledger(stock_account)
  begin
    transactions_csv = %x[hledger aregister -O csv --date2 '#{stock_account}']
    transactions = CSV.parse(transactions_csv, headers:true)
    return DateTime.parse(transactions[-1]['date'])
  rescue
    return DateTime.new()
  end 
end

command :parse_trades do |c|
  c.syntax = 'ib_flex2ledger parse-transactions [options]'
  c.summary = 'Parse trades and output ledger transactions'
  c.description = ''
  c.option '--stock-account STRING', String, 'Ledger account to be used for stock positions'
  c.option '--cash-account STRING', String, 'Ledger account to be used for cash positions'
  c.option '--fees-account STRING', String, 'Ledger account to be used for fees'
  c.option '--dividends-account STRING', String, 'Ledger account to be used for dividends income'
  c.option '--withholdings-account STRING', String, 'Ledger account to be used for withholding tax'
  c.option '--interest-income-account STRING', String, 'Ledger account to be used for interest income'
  c.option '--interest-expense-account STRING', String, 'Ledger account to be used for interest expenses'
  c.option '--ignore-deposits-withdrawals', 'If specified, don\'t output transactions for deposits and withdrawals'
  c.option '--new-only', 'If specified, don\'t output transactions older than the latest recorded ledger transaction on the stock account'
  c.action do |args, options|
    raise ArgumentError.new("--stock-account is required") if options.stock_account.nil?
    options.default :cash_account => options.stock_account,
                    :fees_account => "Expenses:Fees:Brokerage",
                    :dividends_account => "Income:Dividends",
                    :withholdings_account => "Expenses:Taxes:US Withholding Tax",
                    :interest_income_account => "Income:Interest",
                    :interest_expense_account => "Expenses:Interest"

    flex_report = load(args[0])
    parse_trades_from_flex(flex_report, options)
  end
end

command :retrieve_and_parse do |c|
  c.syntax = 'ib_flex2ledger retrieve_and_parse [options]'
  c.summary = 'Retrieve Flex report via API, parse trades from it and output ledger transactions'
  c.description = ''
  c.option '--stock-account STRING', String, 'Ledger account to be used for stock positions'
  c.option '--cash-account STRING', String, 'Ledger account to be used for cash positions'
  c.option '--fees-account STRING', String, 'Ledger account to be used for fees'
  c.option '--dividends-account STRING', String, 'Ledger account to be used for dividends income'
  c.option '--withholdings-account STRING', String, 'Ledger account to be used for withholding tax'
  c.option '--interest-income-account STRING', String, 'Ledger account to be used for interest income'
  c.option '--interest-expense-account STRING', String, 'Ledger account to be used for interest expenses'
  c.option '--ignore-deposits-withdrawals', 'If specified, don\'t output transactions for deposits and withdrawals'
  c.option '--api-token STRING', String, 'Authentication token for the IBKR Flex webservice'
  c.option '--query-id STRING', String, 'Query ID of the Flex query to execute'
  c.option '--wait-seconds INT', Integer, 'Seconds to sleep between executing the query and retrieving the result (default: 5s)'
  c.option '--new-only', 'If specified, don\'t output transactions older than the latest recorded ledger transaction on the stock account'
  c.action do |args, options|
    raise ArgumentError.new("--stock-account is required") if options.stock_account.nil?
    options.default :cash_account => options.stock_account,
                    :fees_account => "Expenses:Fees:Brokerage",
                    :dividends_account => "Income:Dividends",
                    :withholdings_account => "Expenses:Taxes:US Withholding Tax",
                    :interest_income_account => "Income:Interest",
                    :interest_expense_account => "Expenses:Interest",
                    :wait_seconds => 5

    flex = retrieve_flex(options.api_token, options.query_id, options.wait_seconds)
    flex_report = Nokogiri::XML(flex)
    parse_trades_from_flex(flex_report, options)
  end
end

command :print_positions do |c|
  c.syntax = 'ib_flex2ledger print-positions [options]'
  c.summary = 'Print current positions'
  c.description = ''
  c.option '--verbose', 'Include more information than just basic positions'
  c.action do |args, options|
    flex_report = load(args[0])
    statement = flex_report.xpath("//FlexStatement").first
    account_info = statement.xpath("AccountInformation").first
    puts "Positions for #{account_info["name"]}, account #{statement["accountId"]}"
    puts "As per #{statement["toDate"]}"
    puts ""

    assets = statement.xpath("OpenPositions/OpenPosition[@levelOfDetail='SUMMARY']")
    assets.each do |position|
      currency = position["currency"]

      puts "#{position["symbol"]} #{position["position"]}"

      if options.verbose then
        puts " * Percent of NAV: #{position["percentOfNAV"]}%"
        puts " * Cost basis price: #{position["openPrice"]} #{currency}"
        puts " * M2M price: #{position["markPrice"]} #{currency}"
        puts " * Cost basic value : #{position["costBasisMoney"]} #{currency}"
        puts " * M2M value: #{position["positionValue"]} #{currency}"
        puts " * M2M PnL: #{position["fifoPnlUnrealized"]} #{currency}"
        puts ""
      end
    end

    fx = statement.xpath("FxPositions/FxPosition[@levelOfDetail='SUMMARY']")
    fx.each do |position|
      puts "#{position["fxCurrency"]} #{position["quantity"]}"
    end
  end
end

command :dividends_to_csv do |c|
  c.syntax = 'ib_flex2ledger dividends-to-csv [options]'
  c.summary = 'Parse dividends only and output csv'
  c.description = ''
  c.action do |args, options|
    flex_report = load(args[0])
    statement = flex_report.xpath("//FlexStatement").first
    account_info = statement.xpath("AccountInformation").first
    $stderr.puts "Dividends for #{account_info["name"]}, account #{statement["accountId"]}"
    $stderr.puts "Period #{statement["fromDate"]} to #{statement["toDate"]}"

    puts "date,symbol,amount,tax"

    cash_transactions = group_cash_transactions(statement.xpath("CashTransactions/CashTransaction"))
    cash_transactions.sort_by {|key, _| key[:date]}.each do |key, transactions|
      classify_cash_transactions_group(transactions).each do |transaction|
        case transaction[:type]
        when :dividend_with_withholding
          dividend = transaction[:dividend]
          tax = transaction[:tax]

          amount_without_tax = dividend["amount"].to_f + tax["amount"].to_f

          puts "#{dividend["reportDate"]},#{dividend["symbol"]},#{amount_without_tax},#{-tax["amount"].to_f}"
        end
      end
    end
  end
end

command :retrieve_flex do |c|
  c.syntax = 'ib_flex2ledger retrieve-flex [options]'
  c.summary = 'Execute a Flex query via the IBKR Flex webservice and retrieve the result'
  c.description = 'See https://www.interactivebrokers.com/campus/ibkr-api-page/flex-web-service/ for documentation  and setup'
  c.option '--api-token STRING', String, 'Authentication token for the IBKR Flex webservice'
  c.option '--query-id STRING', String, 'Query ID of the Flex query to execute'
  c.option '--wait-seconds INT', Integer, 'Seconds to sleep between executing the query and retrieving the result (default: 5s)'
  c.action do |args, options|
    options.default :wait_seconds => 5

    raise ArgumentError.new("--api-token is required") if options.api_token.nil?
    raise ArgumentError.new("--query-id is required") if options.query_id.nil?

    flex = retrieve_flex(options.api_token, options.query_id, options.wait_seconds)
    puts(flex)
  end
end

File: ib_flex2ledger.py
from lxml import etree
import json
import requests
import sys
from time import sleep
import typer
from datetime import datetime
from collections import defaultdict
from typing import Any
from typing_extensions import Annotated
from dataclasses import dataclass
from functools import reduce
import subprocess
import csv
from io import StringIO


@dataclass(frozen=True)
class Config:
    # Ledger asset account to be used for stock positions
    stock_account: str
    # Ledger asset account to be used for cash positions
    cash_account: str
    # Ledger expenses account to be used for fees
    fees_account: str
    # Ledger revenue account to be used for dividends income
    dividends_account: str
    # Ledger expenses account to be used for withholding tax
    withholdings_account: str
    # Ledger revenue account to be used for interest income
    interest_income_account: str
    # Ledger expenses account to be used for interest expenses
    interest_expense_account: str

    # Authentication token for the IBKR Flex webservice
    api_token: str
    # Query ID of the Flex query to execute
    query_id: str

    @classmethod
    def load(cls, config_file: str):
        config_json = json.load(open(config_file))
        return cls(**config_json)


def transaction_header(trade: dict[str, Any]) -> None:
    print(
        f"{trade.get("settleDateTarget")}={trade.get("tradeDate")} * {trade.get("exchange")}"
    )
    print(f"  ; trade_id: {trade.get("tradeID")}")
    print(f"  ; order_id: {trade.get("ibOrderID")}")


def fees_to_ledger(trade: dict[str, Any], config: Config) -> None:
    print(
        f"  {config.cash_account}  {trade.get("ibCommissionCurrency")} {float(trade.get("ibCommission"))}"
    )
    print(
        f"  {config.fees_account}  {trade.get("ibCommissionCurrency")} {-float(trade.get("ibCommission"))}"
    )


def stock_trade_to_ledger(trade: dict[str, Any], config: Config) -> None:
    transaction_header(trade)
    print(
        f'  {config.stock_account}  "{trade.get("symbol")}" {float(trade.get("quantity"))}'
    )
    print(
        f"  {config.cash_account}  {trade.get("currency")} {float(trade.get("proceeds"))}"
    )
    fees_to_ledger(trade, config)
    print("")


def fx_trade_to_ledger(trade: dict[str, Any], config: Config) -> None:
    base_currency, quote_currency = trade.get("symbol").split(".")

    transaction_header(trade)
    print(f"  {config.cash_account}  {base_currency} {float(trade.get("quantity"))}")
    print(f"  {config.cash_account}  {quote_currency} {float(trade.get("proceeds"))}")
    fees_to_ledger(trade, config)
    print("")


def group_cash_transactions(cash_transactions: list[dict]) -> dict:
    def reduce_fn(hash, transaction):
        key = (transaction.get("conid"), transaction.get("dateTime"))
        if key not in hash:
            hash[key] = [transaction]
        else:
            hash[key].append(transaction)
        return hash

    return reduce(reduce_fn, cash_transactions, {})


def classify_cash_transactions_group(transaction_group):
    grouped = defaultdict(list)
    for transaction in transaction_group:
        grouped[transaction.get("type")].append(transaction)

    results = []

    if len(grouped["Dividends"]) == 1 and len(grouped["Withholding Tax"]) == 1:
        results.append(
            {
                "type": "dividend_with_withholding",
                "dividend": grouped["Dividends"][0],
                "tax": grouped["Withholding Tax"][0],
            }
        )
        del grouped["Dividends"]
        del grouped["Withholding Tax"]

    for key, transactions in grouped.items():
        for transaction in transactions:
            results.append({"type": key, "transaction": transaction})

    return results


app = typer.Typer()


def get_latest_transaction_date_from_hledger(account: str) -> datetime:
    try:
        transactions_csv = subprocess.run(
            ["hledger", "aregister", "-O", "csv", "--date2", account],
            capture_output=True,
            text=True,
        )
        transactions = list(csv.DictReader(StringIO(transactions_csv.stdout)))
        return datetime.fromisoformat(transactions[-1]["date"])
    except Exception:
        return datetime.min


def retrieve_flex(config: Config, wait_seconds: int) -> str:
    print("Executing statement generation...", end="", file=sys.stderr)
    send_request_parameters = {
        "v": 3,
        "t": config.api_token,
        "q": config.query_id,
    }
    send_request_response = requests.get(
        "https://ndcdyn.interactivebrokers.com/AccountManagement/FlexWebService/SendRequest",
        params=send_request_parameters,
    )
    send_request_result = etree.fromstring(send_request_response.text)

    if (
        send_request_result.xpath("//FlexStatementResponse/Status/text()")[0]
        != "Success"
    ):
        raise ValueError(
            f"Response to /SendRequest was not Success:\n{send_request_response.text}"
        )

    reference_code = send_request_result.xpath(
        "//FlexStatementResponse/ReferenceCode/text()"
    )[0]
    print(f"done: {reference_code}", file=sys.stderr)

    print(
        f"Waiting {wait_seconds} seconds for statement to finish generating...",
        end="",
        file=sys.stderr,
    )
    sleep(wait_seconds)
    print("done", file=sys.stderr)

    print("Retrieving generated statement...", end="", file=sys.stderr)
    get_statement_parameters = {
        "v": 3,
        "t": config.api_token,
        "q": reference_code,
    }
    get_statement_response = requests.get(
        "https://ndcdyn.interactivebrokers.com/AccountManagement/FlexWebService/GetStatement",
        params=get_statement_parameters,
    )
    print("done", file=sys.stderr)

    return get_statement_response.text


def parse_trades_from_flex(
    flex_report, new_only: bool, ignore_deposits_withdrawals: bool, config: Config
) -> None:
    statement = flex_report.xpath("//FlexStatement")[0]
    account_info = statement.xpath("AccountInformation")[0]
    print(
        f"Trades for {account_info.get("name")}, account {statement.get("accountId")}",
        file=sys.stderr,
    )
    print(
        f"Period {statement.get("fromDate")} to {statement.get("toDate")}",
        file=sys.stderr,
    )

    if new_only:
        latest_transaction_date = get_latest_transaction_date_from_hledger(
            config.stock_account
        )
        print(
            f"Dropping transactions older than {latest_transaction_date}",
            file=sys.stderr,
        )
    else:
        latest_transaction_date = datetime.min

    trades = statement.xpath("Trades/Trade")
    trades = sorted(
        trades, key=lambda trade: datetime.fromisoformat(trade.get("dateTime"))
    )
    for trade in trades:
        if datetime.fromisoformat(trade.get("tradeDate")) <= latest_transaction_date:
            continue

        match trade.get("assetCategory"):
            case "CASH":
                fx_trade_to_ledger(trade, config)
            case "STK":
                stock_trade_to_ledger(trade, config)
            case _:
                print(
                    f"Dropping trade with unknown assetCategory={trade.get("assetCategory")}",
                    file=sys.stderr,
                )
                print(trade, file=sys.stderr)

    cash_transactions = group_cash_transactions(
        statement.xpath("CashTransactions/CashTransaction[@levelOfDetail='DETAIL']")
    )
    cash_transactions = sorted(cash_transactions.items(), key=lambda item: item[0][1])
    for key, transactions in cash_transactions:
        if datetime.fromisoformat(key[1]) <= latest_transaction_date:
            continue

        grouped_transactions = classify_cash_transactions_group(transactions)
        for transaction in grouped_transactions:
            match transaction["type"]:
                case "dividend_with_withholding":
                    dividend = transaction["dividend"]
                    tax = transaction["tax"]

                    print(f"{dividend.get("reportDate")} * {dividend.get("symbol")}")
                    print(f"  ; {dividend.get("description")}")
                    print(
                        f"  {config.dividends_account}  {dividend.get("currency")} {-float(dividend.get("amount"))}"
                    )
                    print(
                        f"  {config.withholdings_account}  {tax.get("currency")} {-float(tax.get("amount"))}"
                    )
                    print(f"  {config.cash_account}")
                    print(f"")
                case "Dividends":
                    dividend = transaction["transaction"]

                    print(f"{dividend.get("reportDate")} * {dividend.get("symbol")}")
                    print(f"  ; {dividend.get("description")}")
                    print(
                        f"  {config.dividends_account}  {dividend.get("currency")} {-float(dividend.get("amount"))}"
                    )
                    print(f"  {config.cash_account}")
                    print(f"")
                case "Broker Interest Received":
                    transaction = transaction["transaction"]

                    print(f"{transaction.get("reportDate")} * Interactive Brokers")
                    print(f"  ; {transaction.get("description")}")
                    print(
                        f"  {config.interest_income_account}  {transaction.get("currency")} {-float(transaction.get("amount"))}"
                    )
                    print(f"  {config.cash_account}")
                    print(f"")
                case "Broker Interest Paid":
                    transaction = transaction["transaction"]

                    print(f"{transaction.get("reportDate")} * Interactive Brokers")
                    print(f"  ; {transaction.get("description")}")
                    print(
                        f"  {config.interest_expense_account}  {transaction.get("currency")} {-float(transaction.get("amount"))}"
                    )
                    print(f"  {config.cash_account}")
                    print(f"")
                case "Other Fees":
                    transaction = transaction["transaction"]

                    print(f"{transaction.get("reportDate")} * Interactive Brokers")
                    print(f"  ; {transaction.get("description")}")
                    print(
                        f"  {config.fees_account}  {transaction.get("currency")} {-float(transaction.get("amount"))}"
                    )
                    print(f"  {config.cash_account}")
                    print(f"")
                case "Deposits/Withdrawals":
                    if ignore_deposits_withdrawals:
                        continue
                    transaction = transaction["transaction"]

                    print(f"{transaction.get("reportDate")} * UNKNOWN")
                    print(f"  ; {transaction.get("description")}")
                    print(
                        f"  {config.cash_account}  {transaction.get("currency")} {float(transaction.get("amount"))}"
                    )
                    print(f"  UNKNOWN_ACCOUNT")
                    print(f"")
                case _:
                    transaction = transaction["transaction"]
                    amount = float(transaction.get("amount"))

                    print(f"{transaction.get("reportDate")} * Interactive Brokers")
                    print(f"  ; {transaction.get("description")}")
                    print(f"  ; cash_transaction_type: {transaction.get("type")}")
                    print(
                        f"  {config.cash_account}  {transaction.get("currency")} {amount}"
                    )
                    if amount < 0:
                        print(f"  {config.fees_account}")
                    else:
                        print(f"  UNKNOWN_ACCOUNT")
                    print(f"")


@app.command("retrieve-flex")
def retrieve_flex_command(
    config_file: Annotated[str, typer.Option()], wait_seconds: int = 5
) -> None:
    config = Config.load(config_file)
    print(retrieve_flex(config, wait_seconds))


@app.command("parse-trades")
def parse_trades_command(
    flex_file: str,
    config_file: Annotated[str, typer.Option()],
    new_only: bool = False,
    ignore_deposits_withdrawals: bool = False,
) -> None:
    config = Config.load(config_file)
    with open(flex_file, "rb") as f:
        flex_report = etree.parse(f).getroot()
        parse_trades_from_flex(
            flex_report, new_only, ignore_deposits_withdrawals, config
        )


if __name__ == "__main__":
    app()
